radar/src/radar.rs

```diff
@@ -0,0 +1,603 @@
+    current_position: &Position,
+    radar_receiver: &RadarReceiver,
+    old_contacts: &HashMap<String, RadarContact>,
+    all_positions: &HashMap<String, Position>,
+) -> Vec<RadarContactDelta> {
+    let mut deltas: Vec<RadarContactDelta> = Vec::new();
+    let contacts: Vec<String> = old_contacts
+        .values()
+        .map(|rc| rc.clone().entity_id)
+        .collect();
+    for (k, v) in all_positions.iter() {
+        if contacts.contains(k) {
+            let mut rid: String = "".to_string();
+            for (key, val) in old_contacts.iter() {
+                if val.entity_id == *k {
+                    rid = key.to_string();
+                    break;
+                }
+            }
+            if within_radius(current_position, v, radar_receiver.radius) {
+                let vector_to = current_position.vector_to(v);
+                deltas.push(RadarContactDelta::Change(
+                    rid.replace(":", "."),
+                    RadarContact {
+                        entity_id: k.clone().to_string(),
+                        distance: vector_to.mag,
+                        heading_xy: vector_to.heading_xy,
+                        heading_z: vector_to.heading_z,
+                    },
+                ));
+            } else {
+                deltas.push(RadarContactDelta::Remove(rid));
+            }
+        } else if entity_id != k && within_radius(current_position, &v, radar_receiver.radius) {
+            let vector_to = current_position.vector_to(v);
+            deltas.push(RadarContactDelta::Add(RadarContact {
+                entity_id: k.clone().to_string(),
+                distance: vector_to.mag,
+                heading_xy: vector_to.heading_xy,
+                heading_z: vector_to.heading_z,
+            }));
+        }
+    }
+    deltas
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+enum RadarContactDelta {
+    Add(RadarContact),
+    Remove(String),
+    Change(String, RadarContact),
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct PositionValue {
+    values: Position,
+}
+
+/// Receives messages on the subject `event.decs.components.{shard}.{entity}.position.change`
+/// Stores entity position in-memory in the POSITIONS HashMap
+/// The cache is used later to discover nearby radar_contacts
+pub(crate) fn handle_entity_position_change(
+    _ctx: &CapabilitiesContext,
+    msg: messaging::BrokerMessage,
+) -> CallResult {
+    let subject: Vec<&str> = msg.subject.split('.').collect();
+    let position_value: PositionValue = serde_json::from_slice(&msg.body)?;
+    let position: Position = position_value.values;
+    POSITIONS
+        .write()
+        .unwrap()
+        .insert(subject[4].to_string(), position);
+    Ok(vec![])
+}
+
+/// Helper function to clean up determining if an entity is within a radius
+fn within_radius(entity: &Position, target: &Position, radius: f64) -> bool {
+    entity.distance_to(target) <= radius
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
This test mod is LONG. Is that an issue? Maybe refactoring could help?

Reply:
One thought here might be to store test data in some JSON files and then your individual tests can load a specific use case and assert the expected values (which could also be stored in files)